<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Комментарий ревьюера</b>


Привет, Александр! Давай знакомиться! Меня зовут Дмитрий Махортов, и я буду проверять твой проект. Сразу предлагаю общение на "ты" 🙂, но если тебе это не комфортно, то дай знать, и мы перейдем на "вы". 

Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе погрузиться в увлекательный мир работы с данными и вырасти в крепкого профи. Это отдаленная цель. А ближайшая - сделать твою работу еще лучше )).


Вижу, что ты очень старался, делая проект. Все ключевые этапы в работе выполнены, и в целом с проектом ты справился. 
    
Расскажу как обычно проходит проверка:    
Бывают моменты, которые требуют пристального внимания. Комментарии по ним выделены <span style='background-color:#F7B3A4'> красным цветом </span> и обозначены значком 🛑. После их доработки проект будет принят.  🙂

<span style='background-color:#F9EDA6'>Жёлтым цветом </span> и значком ⚠️ выделено то, что в следующий раз можно сделать по-другому. Можно учесть эти комментарии при выполнении будущих заданий или доработать текущий проект (однако это не обязательно). 

<span style='background-color:#B7EBA7'> Зеленым цветом </span> и значком ✅ отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах. Или советы "со звездочкой", которые помогут в будущем.

    
Давай работать над проектом в диалоге: **если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом**. Выбери для своих комментариев какой-то заметный цвет, так мне будет легче отследить изменения. Пожалуйста, **не перемещай, не изменяй и не удаляй мои комментарии**. Всё это поможет выполнить повторную проверку твоего проекта оперативнее».
</div>

# Краткое описание проекта: рекомендация тарифов

Мегалайн желаем построить систему, способную проанализировать поведение клиентов и предложить пользователям новый тариф: «Смарт» или «Ультра». В нашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не требуется.

Необходимо построить модель с максимально большим значением *accuracy*. Нужно довести долю правильных ответов по крайней мере до 0.75, а также проверить *accuracy* на тестовой выборке.

# Описание данных
 - сalls — количество звонков,
 - minutes — суммарная длительность звонков в минутах,
 - messages — количество sms-сообщений,
 - mb_used — израсходованный интернет-трафик в Мб,
 - is_ultra — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0).

# план работы: 
1. [Открыть и изучить файл](#1_plan)

2. [Разбить данные на выборки](#2_plan)

3. [Исследовать модели](#3_plan)

4. [проверить лучшую модель на тестовой выборке](#4_plan)


<a id="1_plan"></a> 
## Откройте и изучите файл

**1.1 импортируем необходимые библиотеки**

In [1]:
import pandas as pd
import numpy as np
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Отлично, все нужные библиотеки импортированы в начале ноутбука.Это хорошая практика.</div>

In [2]:
df = pd.read_csv('/datasets/users_behavior.csv')   # читаем df
df.head()


calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

In [3]:
df.info()
# определим тип данных и возможные пропуски. Все в порядке, пропусков нет

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


In [4]:
df.is_ultra.value_counts() # необходимо понять сколько абонентов Ультра и Смарт на данный момент
# исходя из данных, видно, что абонентов Ультра меньше чем Смарт примерно в 2 раза

0    2229
1     985
Name: is_ultra, dtype: int64

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Данные загружены и исследованы.</div>

<a id="2_plan"></a> 
## Разбейте данные на выборки

# 2.1 извлечем из данных признаки и целевой признак
 - признаки: calls, minutes, messeges, mb_used
 - целевой признак: is_ultra

In [5]:
features = df.drop('is_ultra', axis=1)
target = df['is_ultra']

# 2.2 разделим данные на train, test, valid выборки. 60% на обучающую, по 20% на тестовую и валидационную

In [6]:
X, X_test, y, y_test = train_test_split(features, target, test_size=0.2, random_state=12345)
X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size=0.25, random_state=12345)

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Молодец, корректно разделил данные на три выборки.
На будущее, обрати внимание на аргумент stratify функции train_test_split. Стратификация особенно важна, когда у нас мало данных и возможен дисбаланс классов в обучении / валидации / тесте.
</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Лучше явно вывести размеры получившихся выборок (с использованием .shape и print), это поможет быстро оценить корректность разделения на выборки.
</div>

<a id="3_plan"></a> 
## Исследуйте модели

# Начнем с Решающего дерева

In [7]:
best_accuracy = 0.75  # согласно условиям задания
best_depth = 0
# в best_model будем хранить лучшие параметры, в best_accuracy супер точность, в best_depth максимальную глубину дерева

In [8]:
for depth in range(1,6):
    model_tree = DecisionTreeClassifier(random_state = 12345, max_depth = depth) # инициализировали модель 
#DecisionTreeClassifier с параметром random_state=12345 и max_depth=depth
    model_tree.fit(X_train, y_train) # обучаем модель на тренировочной выборке
    predictions = model_tree.predict(X_valid) # получаем предсказания модели на валидационной выборке
    accuracy = accuracy_score(y_valid, predictions) # вычсиляем точность 
    if accuracy > best_accuracy:
                    best_depth = depth
                    best_accuracy = accuracy

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Решающее дерево исследовано корректно:
    
 - исследовано 5 значений гиперпараметров
 - модель обучена на обучающем наборе
 - получена оценка качества на валидационном наборе
 - перебор гиперпараметров осуществляется в цикле
</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Рекомендую посмотреть как при изменении гиперпараметров меняется точность на обучающем и валидационых наборах. Для этого с помощью print() выводи эти метрики на каждом шагу цикла. Также уместно использовать графики.
</div>

In [9]:
display("супер точность:", best_accuracy, "Глубина дерева:", best_depth)

'супер точность:'

0.7651632970451011

'Глубина дерева:'

3

# вывод по модели Решающее дерево
1. Качество модели (Accuracy = 0.76). Модель решающего дерева удовлетворяет условиям задания

# 3.2 Следующая модель Случайный лес

In [10]:
best_accuracy = 0.75
best_est = 0
best_depth = 0
for est in range(10, 51, 10): # с количеством деревьев: от 10 до 50 с шагом 10
    for depth in range (1, 11): # с максимальной глубиной от 1 до 10
        model_forest = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth=depth) # инициализировали модель RandomForestRegressor с параметрами random_state=12345, n_estimators и max_depth
        model_forest.fit(X_train, y_train) #обучение модели на тренировочной выборке
        predictions = model_forest.predict(X_valid) # получаем предсказания модели на валидационной выборке
        accuracy = accuracy_score(y_valid, predictions)
        if accuracy > best_accuracy:
                    best_depth = depth
                    best_est = est
                    best_accuracy = accuracy

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Решающее дерево исследовано корректно:
    
 - исследовано 5х10 значений гиперпараметров
 - модель обучена на обучающем наборе
 - получена оценка качества на валидационном наборе
 - перебор гиперпараметров осуществляется в цикле
</div>

In [11]:
display("супер точность:", best_accuracy, "Количество деревьев:", best_est, "Максимальная глубина:", best_depth)

'супер точность:'

0.7978227060653188

'Количество деревьев:'

50

'Максимальная глубина:'

10

# вывод поисследованию модели Случайный лес
1. Точность в модели случайного леса выше более чем на 3% по сравнению с решающим деревом

# 3.3 Следующая модель Логистическая регрессия

In [12]:
best_accuracy = 0.75

In [13]:

model_log = LogisticRegression(random_state=12345, solver='lbfgs', max_iter=1000) # инициализиировали модель с гиперпараметрами olver='lbfgs', max_iter=1000
model_log.fit(X_train, y_train)  # обучаем модель на тренировочной выборке
#result = model_log.score(X_valid, y_valid)
predictions = model_log.predict(X_valid) # ищем предсказания модели на валидационной выбоке
accuracy = accuracy_score(y_valid, predictions)
display("Accuracy модели логистической регрессии на валидационной выборке:", accuracy)


'Accuracy модели логистической регрессии на валидационной выборке:'

0.7262830482115086

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
С логистической регрессией тоже все корректно
</div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Рассмотрены все три изученных алгоритма (решающее дерево, случайный лес, логистическая регрессия)
</div>


# Вывод
1. Модель Случайного леса по критерию accuracy показывает наилучшую точность 

2. Наилучшая точность: 0.79

3. гиперпараметры модели: n_estimators = 50, max_depth = 10

3. Логистическая регрессия даёт точность 0,72, что ниже на 6%, чем случайный лес

4. Решающее дерево дает точноть 0,76, что ниже, чем Случайный лес на 3%

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Нужно сделать вывод, какая модель лучшая по по результатам валидации. И явно описать не только алгоритм, но и все найденные гиперпараметры.
</div>

---

<div class="alert alert-info">
<h2> Комментарий от студента </h2>    
вывод дополнил
</div>

---

<a id="4_plan"></a> 
## Проверьте модель на тестовой выборке

4.1 наилучшей точности удалось достичь с моделью Случайного леса, наилучшая точность: 0.7978227060653188, поэтому на тестовой выборке будем проверять именно эту модель

In [14]:
model_forest = RandomForestClassifier(random_state=12345, n_estimators=50, max_depth=10)
model_forest.fit(X, y)
predictions = model_forest.predict(X_test)
accuracy = accuracy_score(y_test, predictions)
display("Точность модели на тестовой выборке:", accuracy)


'Точность модели на тестовой выборке:'

0.7947122861586314

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Выводы стоит описывать в markdown ячейках, а не в комментариях. 
</div>

---

<div class="alert alert-info">
<h2> Комментарий от студента </h2>    
поправил
</div>

---

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Молодец: 
    
  - выбрана лучшая модель
  - тестирование произведено корректно 
  - достигнута хорошая точность на тестовой выборке
</div>


<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Отлично! Очень хорошо,что ты переобучил модель сумме обучающей и валидационных выборок. 
</div>


# Общий вывод сравнение моделей
1. проведен анализ трех моделей: Решающее дерево, Случайный лес, Логистическая регрессия

2. Каждая модель были использованы различные гиперпараметры: max_depth=max_depth, n_estimators=est, solver='lbfgs', max_iter=1000

3. супер точность (accuracy) была достигнута для модели Случайного леса:
   Accuracy на валидационной выборке: 0.7978227060653188
   Accuracy на тестовой выборке: 0.7947122861586314

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
👍 </div>

## (бонус) Проверьте модели на адекватность

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
В этом задании предполагался достаточно простой подход. Нужно измерить точность для "наивного" алгоритма, например предсказывающего случайный ответ или константный. Рекомендую посмотреть, какую  точность дает "наивный" алгоритм, всегда предсказывающий самый популярный ответ (у нас это 0). Вычислить это можно вручную или используя DummyClassifier.
</div>

In [15]:
from sklearn.dummy import DummyClassifier #импортируем библиотеку DummyClassifier для определения модели на адекватность
dummy_model = DummyClassifier(strategy="most_frequent", random_state=0) #  возьмем самый популярный ответ 0
dummy_model.fit(X_train, y_train) # обучаем модель на тренировочной выборке
dummy_model.score(X_train, y_train) # считаем точность наивного алгорита 

0.6945020746887967

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
<b>v2</b> 👍 </div>

# Вывод
1. наивный алгоритм (видимо простейгая модель) ошибается в 30,5%
2. Случайный лес, попавший в тестовую выборку и считающуюся наиболее удачной показывает точность сильно выше = 0,79% и ошибается она явно меньше = 21%
3. Выбранная нами модель Случайный лес лучше, чем наивная модель, и она адекватна.

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x] Jupyter Notebook открыт
- [x] Весь код исполняется без ошибок
- [x] Ячейки с кодом расположены в порядке исполнения
- [x] Выполнено задание 1: данные загружены и изучены
- [x] Выполнено задание 2: данные разбиты на три выборки
- [x] Выполнено задание 3: проведено исследование моделей
    - [x] Рассмотрено больше одной модели
    - [x] Рассмотрено хотя бы 3 значения гипепараметров для какой-нибудь модели
    - [x] Написаны выводы по результатам исследования
- [x] Выполнено задание 3: Проведено тестирование
- [x] Удалось достичь accuracy не меньше 0.75


<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Заключительный комментарий</b>

Все хорошо: твоя работа выполнена верно, без ошибок,  с минимальными помарками.
Также работа выглядит аккуратной и хорошо оформленной.
Вижу, что тебе дается python, и инструментарий датасайнтиста.

К твоей работе нет существенных замечаний, все обязательные этапы ты выполнил. Я готов принять проект. Возвращаю тебе его чтобы ты мог ознакомиться с моими комментариями, возможно у тебя появятся вопросы.

Жду работу на финальное ревью ))</div>

---

<div class="alert alert-info">
<h2> Комментарий от студента </h2>    
Привет, Дмитрий! Спасибо за твои замечания и подсказки!) Я исправил по ходу проекта некотторые вещи:
    1. все выводы и свои мысли переместил в ячейки markdown
    2. поправил вывод по Случайному лесу
    3. выполнил бонусное задание, при помощи DummyClassifier
</div>

---

<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Заключительный комментарий v2</b>

Привет! Да, вижу что ты сделал допонительную работу по моим комментариям, молодец!

Проект принят, поздравляю! И желаю дальнейших успехов!</div>